File: cpp/src/common/status.cpp
// Copyright 2023 Zilliz
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "milvus-storage/common/status.h"
#include <cstdio>
#include <string>
namespace milvus_storage {

Status& Status::operator=(const Status& s) {
  code_ = s.code_;
  msg_ = s.msg_;
  return *this;
}

std::string Status::ToString() const {
  char tmp[30];
  std::string res;
  switch (code_) {
    case kOk:
      return "OK";
      break;
    case kArrowError:
      res = "ArrowError: ";
      break;
    case kInvalidArgument:
      res = "InvalidArgument: ";
      break;
    case kInternalStateError:
      res = "InternalStateError: ";
      break;
    case kFileNotFound:
      res = "FileNotFound: ";
    default:
      std::sprintf(tmp, "Unknown code(%d): ", code_);
      res = tmp;
      break;
  }
  res.append(msg_);
  return res;
}

}  // namespace milvus_storage

File: go/common/status/status.go
package status

type Code int32

const (
	KOk                 Code = 0
	kArrowError         Code = 1
	kInvalidArgument    Code = 2
	kInternalStateError Code = 3
)

type Status struct {
	code Code
	msg  string
}

func NewStatus(code Code, msg string) *Status {
	return &Status{
		code: code,
		msg:  msg,
	}
}

func (s *Status) Code() Code {
	return s.code
}

func (s *Status) Msg() string {
	return s.msg
}

func OK() Status {
	return Status{
		code: KOk,
	}
}

func ArrowError(msg string) Status {
	return Status{
		code: kArrowError,
		msg:  msg,
	}
}

func InvalidArgument(msg string) Status {
	return Status{
		code: kInvalidArgument,
		msg:  msg,
	}
}

func InternalStateError(msg string) Status {
	return Status{
		code: kInternalStateError,
		msg:  msg,
	}
}

func (s *Status) IsOK() bool {
	return s.code == KOk
}

func (s *Status) IsArrowError() bool {
	return s.code == kArrowError
}

func (s *Status) IsInvalidArgument() bool {
	return s.code == kInvalidArgument
}

func (s *Status) IsInternalStateError() bool {
	return s.code == kInternalStateError
}
